File: frontend/src/utils/overallRating.ts
import type { EvaluationZone } from "./evaluation";

export interface SignalItem {
  category: string;
  label: string;
  impact: number;
}

export interface OverallRatingResult {
  label: string;
  score: number;
  color: string;
  bgColor: string;
  signals: SignalItem[];
}

interface ComputeInputs {
  evaluationZone: EvaluationZone | null;
  growthRateLabel: string | null;
  roeTrend: "improving" | "declining" | "stable" | null;
  epsCagr3y: number | null;
  epsGrowthYoy: number | null;
  slRatio: number | null;
  momentumSignal?: "strong_buy" | "buy" | "neutral" | "caution" | "sell" | null;
  dividendYield?: number | null;
  payoutRatio?: number | null;
  consecutiveDividendYears?: number | null;
  progressiveDividendYears?: number | null;
  fcfYield?: number | null;
  fcfMargin?: number | null;
  fcf?: number | null;
  prevFcf?: number | null;
}

export function computeOverallRating(inputs: ComputeInputs): OverallRatingResult {
  const signals: SignalItem[] = [];
  let score = 0;

  // --- バリュエーション ---
  const valuation = (() => {
    switch (inputs.evaluationZone) {
      case "very_cheap":  return { impact: +3, label: "超割安（30%以上割安）" };
      case "cheap":       return { impact: +1, label: "割安（10〜30%割安）" };
      case "fair":        return { impact:  0, label: "適正（±10%）" };
      case "expensive":   return { impact: -2, label: "割高（10〜30%割高）" };
      case "very_expensive": return { impact: -4, label: "危険（30%以上割高）" };
      default:            return { impact:  0, label: "データなし" };
    }
  })();
  score += valuation.impact;
  signals.push({ category: "評価ゾーン", label: valuation.label, impact: valuation.impact });

  // --- 成長率評価 ---
  const growth = (() => {
    switch (inputs.growthRateLabel) {
      case "かなり強気":       return { impact: +2, label: "かなり強気（市場折込成長率≥7%）" };
      case "非常に優秀":
      case "優秀":             return { impact: +1, label: `${inputs.growthRateLabel}（市場折込成長率≥3%）` };
      case "普通":
      case "低成長":           return { impact:  0, label: `${inputs.growthRateLabel}（市場折込成長率≥0%）` };
      case "マイナス成長":     return { impact: -1, label: "マイナス成長（市場折込成長率<0%）" };
      case "ROE持続性に疑念":  return { impact: -2, label: "ROE持続性に疑念（PBR<1, 市場が収益力を不信視）" };
      default:                 return { impact:  0, label: "データなし" };
    }
  })();
  score += growth.impact;
  signals.push({ category: "成長率評価", label: growth.label, impact: growth.impact });

  // --- ROEトレンド ---
  const roeTrendImpact = (() => {
    switch (inputs.roeTrend) {
      case "improving": return { impact: +1, label: "3期連続改善" };
      case "declining": return { impact: -1, label: "3期連続悪化" };
      case "stable":    return { impact:  0, label: "横ばい" };
      default:          return { impact:  0, label: "データなし" };
    }
  })();
  score += roeTrendImpact.impact;
  signals.push({ category: "ROEトレンド", label: roeTrendImpact.label, impact: roeTrendImpact.impact });

  // --- EPS成長（3年CAGRを優先、なければYoY）---
  const epsGrowth = inputs.epsCagr3y ?? inputs.epsGrowthYoy;
  const epsImpact = (() => {
    if (epsGrowth == null) return { impact: 0, label: "データなし" };
    const source = inputs.epsCagr3y != null ? "CAGR3年" : "前年比";
    const pct = (epsGrowth * 100).toFixed(1);
    if (epsGrowth >= 0.2) return { impact: +1, label: `${pct}%（${source}、高成長）` };
    if (epsGrowth <= -0.2) return { impact: -1, label: `${pct}%（${source}、大幅減益）` };
    return { impact: 0, label: `${pct}%（${source}、中程度）` };
  })();
  score += epsImpact.impact;
  signals.push({ category: "EPS成長", label: epsImpact.label, impact: epsImpact.impact });

  // --- 信用残高（信売比率）---
  const marginImpact = (() => {
    if (inputs.slRatio == null) return { impact: 0, label: "データなし" };
    const r = inputs.slRatio;
    const fmt = r.toFixed(2);
    if (r < 0.05) return { impact: -4, label: `信売比率 ${fmt} — 極端な買い偏り（強制売りリスク大）` };
    if (r < 0.30) return { impact: -1, label: `信売比率 ${fmt} — 買い偏り` };
    if (r <= 2.00) return { impact:  0, label: `信売比率 ${fmt} — 均衡` };
    return { impact: 0, label: `信売比率 ${fmt} — 売り偏り` };
  })();
  score += marginImpact.impact;
  signals.push({ category: "信用残高", label: marginImpact.label, impact: marginImpact.impact });

  // --- モメンタム（52週高値効果）---
  const momentumImpact = (() => {
    switch (inputs.momentumSignal) {
      case "strong_buy": return { impact: +2, label: "52w高値近接 + 出来高急増" };
      case "buy":        return { impact: +1, label: "上昇トレンド確認" };
      case "neutral":    return { impact:  0, label: "中間圏" };
      case "caution":    return { impact: -1, label: "下落トレンド" };
      case "sell":       return { impact: -1, label: "底値圏" };
      default:           return { impact:  0, label: "データなし" };
    }
  })();
  score += momentumImpact.impact;
  signals.push({ category: "モメンタム", label: momentumImpact.label, impact: momentumImpact.impact });

  // --- 配当評価 ---
  const dividendImpact = (() => {
    const dy = inputs.dividendYield;
    const pr = inputs.payoutRatio;
    const consec = inputs.consecutiveDividendYears;
    const prog = inputs.progressiveDividendYears;

    if (dy == null || dy === 0) {
      if (dy === 0) return { impact: -1, label: "無配" };
      return { impact: 0, label: "データなし" };
    }
    const highYield = dy >= 3.5;
    const riskyPayout = pr != null && pr > 70;
    const healthyPayout = pr == null || pr <= 60;

    if (highYield && riskyPayout) {
      if (prog == null || prog === 0) {
        return { impact: -2, label: `利回り${dy.toFixed(1)}% / 性向${pr!.toFixed(0)}% + 累進なし` };
      }
      return { impact: -1, label: `利回り${dy.toFixed(1)}% / 性向${pr!.toFixed(0)}%（持続性リスク）` };
    }
    if (highYield && healthyPayout) {
      if (prog != null && prog >= 3) {
        return { impact: +2, label: `利回り${dy.toFixed(1)}% / 性向${pr?.toFixed(0) ?? "N/A"}% / 累進${prog}年` };
      }
      return { impact: +1, label: `高配当${dy.toFixed(1)}% + 健全性向` };
    }
    if (consec != null && consec >= 5) {
      return { impact: +1, label: `連続配当${consec}年` };
    }
    return { impact: 0, label: `配当あり（利回り${dy.toFixed(1)}%）` };
  })();
  score += dividendImpact.impact;
  signals.push({ category: "配当評価", label: dividendImpact.label, impact: dividendImpact.impact });

  // --- FCF評価 ---
  const fcfImpact = (() => {
    const fcf = inputs.fcf;
    const fy = inputs.fcfYield;
    const fm = inputs.fcfMargin;

    if (fcf == null) return { impact: 0, label: "データなし" };

    if (fcf < 0) {
      if (inputs.prevFcf != null && inputs.prevFcf < 0) {
        return { impact: -2, label: "FCF 2期連続マイナス" };
      }
      return { impact: -1, label: "FCFマイナス" };
    }

    if (fy != null && fy >= 8 && fm != null && fm >= 10) {
      return { impact: +2, label: `FCF利回り${fy.toFixed(1)}% + マージン${fm.toFixed(0)}%` };
    }
    if (fy != null && fy >= 5) {
      return { impact: +1, label: `FCF利回り${fy.toFixed(1)}%` };
    }
    return { impact: 0, label: `FCFプラス${fy != null ? `（利回り${fy.toFixed(1)}%）` : ""}` };
  })();
  score += fcfImpact.impact;
  signals.push({ category: "FCF評価", label: fcfImpact.label, impact: fcfImpact.impact });

  // --- ラベル決定 ---
  const rating = (() => {
    if (score >= 6) return { label: "超割安",   color: "#1b5e20", bgColor: "#e8f5e9" };
    if (score >= 2) return { label: "買い推奨", color: "#2e7d32", bgColor: "#c8e6c9" };
    if (score >= -1) return { label: "レンジ中", color: "#f57f17", bgColor: "#fff9c4" };
    if (score >= -5) return { label: "下落警戒", color: "#e65100", bgColor: "#ffe0b2" };
    return             { label: "購入危険",   color: "#b71c1c", bgColor: "#ffcdd2" };
  })();

  return { ...rating, score, signals };
}

File: backend/apps/stocks/domain/overall_rating.py
"""総合評価スコアの算出（フロント frontend/src/utils/overallRating.ts の Python 移植）。

8カテゴリ（評価ゾーン / 成長率評価 / ROEトレンド / EPS成長 / 信売比率 / モメンタム /
配当 / FCF）の impact を合算してスコアを返す。ラベル・色は表示専用のためフロントに残し、
ここでは数値スコアのみを算出する（一覧のソート・フィルタに使う）。

TS 版との数値一致は golden vector テスト（test_overall_rating.py と
frontend の overallRating.test.ts で同一入力→同一 score）で担保する。
"""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal


@dataclass
class OverallRatingInputs:
    """総合評価スコアの入力。値が無い項目は None（＝データなし扱いで impact 0）。

    比率系（eps_cagr_3y / eps_growth_yoy）は比率スケール（0.2 = 20%）。
    %スケール系（dividend_yield / payout_ratio / fcf_yield / fcf_margin）は
    フロントと同じく既に % 値（3.5 = 3.5%）で渡す。
    """

    evaluation_zone: str | None
    growth_rate_label: str | None
    roe_trend: str | None
    eps_cagr_3y: Decimal | None
    eps_growth_yoy: Decimal | None
    sl_ratio: Decimal | None
    momentum_signal: str | None
    dividend_yield: Decimal | None
    payout_ratio: Decimal | None
    consecutive_dividend_years: int | None
    progressive_dividend_years: int | None
    fcf_yield: Decimal | None
    fcf_margin: Decimal | None
    fcf: int | None
    prev_fcf: int | None


def _valuation_impact(zone: str | None) -> int:
    return {
        "very_cheap": 3,
        "cheap": 1,
        "fair": 0,
        "expensive": -2,
        "very_expensive": -4,
    }.get(zone or "", 0)


def _growth_impact(label: str | None) -> int:
    if label == "かなり強気":
        return 2
    if label in ("非常に優秀", "優秀"):
        return 1
    if label in ("普通", "低成長"):
        return 0
    if label == "マイナス成長":
        return -1
    if label == "ROE持続性に疑念":
        return -2
    return 0


def _roe_trend_impact(trend: str | None) -> int:
    return {"improving": 1, "declining": -1, "stable": 0}.get(trend or "", 0)


def _eps_growth_impact(cagr3y: Decimal | None, yoy: Decimal | None) -> int:
    eps_growth = cagr3y if cagr3y is not None else yoy
    if eps_growth is None:
        return 0
    if eps_growth >= Decimal("0.2"):
        return 1
    if eps_growth <= Decimal("-0.2"):
        return -1
    return 0


def _margin_impact(sl_ratio: Decimal | None) -> int:
    if sl_ratio is None:
        return 0
    if sl_ratio < Decimal("0.05"):
        return -4
    if sl_ratio < Decimal("0.30"):
        return -1
    # 0.05〜2.00 は均衡(0)、2.00超も 0（TS版と一致）
    return 0


def _momentum_impact(signal: str | None) -> int:
    return {
        "strong_buy": 2,
        "buy": 1,
        "neutral": 0,
        "caution": -1,
        "sell": -1,
    }.get(signal or "", 0)


def _dividend_impact(
    dy: Decimal | None,
    pr: Decimal | None,
    consec: int | None,
    prog: int | None,
) -> int:
    if dy is None or dy == 0:
        if dy == 0:
            return -1  # 無配
        return 0  # データなし
    high_yield = dy >= Decimal("3.5")
    risky_payout = pr is not None and pr > Decimal("70")
    healthy_payout = pr is None or pr <= Decimal("60")

    if high_yield and risky_payout:
        if prog is None or prog == 0:
            return -2
        return -1
    if high_yield and healthy_payout:
        if prog is not None and prog >= 3:
            return 2
        return 1
    if consec is not None and consec >= 5:
        return 1
    return 0


def _fcf_impact(
    fcf: int | None,
    fcf_yield: Decimal | None,
    fcf_margin: Decimal | None,
    prev_fcf: int | None,
) -> int:
    if fcf is None:
        return 0
    if fcf < 0:
        if prev_fcf is not None and prev_fcf < 0:
            return -2
        return -1
    if fcf_yield is not None and fcf_yield >= Decimal("8") and fcf_margin is not None and fcf_margin >= Decimal("10"):
        return 2
    if fcf_yield is not None and fcf_yield >= Decimal("5"):
        return 1
    return 0


def compute_overall_score(inputs: OverallRatingInputs) -> int:
    """8カテゴリの impact を合算した総合評価スコアを返す。"""
    return (
        _valuation_impact(inputs.evaluation_zone)
        + _growth_impact(inputs.growth_rate_label)
        + _roe_trend_impact(inputs.roe_trend)
        + _eps_growth_impact(inputs.eps_cagr_3y, inputs.eps_growth_yoy)
        + _margin_impact(inputs.sl_ratio)
        + _momentum_impact(inputs.momentum_signal)
        + _dividend_impact(
            inputs.dividend_yield,
            inputs.payout_ratio,
            inputs.consecutive_dividend_years,
            inputs.progressive_dividend_years,
        )
        + _fcf_impact(inputs.fcf, inputs.fcf_yield, inputs.fcf_margin, inputs.prev_fcf)
    )
